File: lib/data/airports.dart
// =============================================================================
// BRE4CH - Regional Airport Database
// Verified ICAO/IATA codes, coordinates, and FlightRadar24 links
// Countries: UAE, KSA, Oman, Qatar, Bahrain, Israel, Lebanon
//
// Sources:
// - ICAO Doc 7910 (Location Indicators)
// - OurAirports (CC-BY-SA, ourairports.com)
// - FlightRadar24 (public airport pages)
// =============================================================================

class AirportData {
  final String icao;
  final String iata;
  final String name;
  final String city;
  final String country;
  final String countryCode;
  final String flag;
  final double lat;
  final double lng;
  final bool isMilitary;
  final String fr24Url;        // FlightRadar24 airport page
  final String notamUrl;       // ICAO NOTAM source

  const AirportData({
    required this.icao,
    required this.iata,
    required this.name,
    required this.city,
    required this.country,
    required this.countryCode,
    required this.flag,
    required this.lat,
    required this.lng,
    this.isMilitary = false,
    required this.fr24Url,
    required this.notamUrl,
  });
}

const List<AirportData> regionalAirports = [
  // ── UAE ──────────────────────────────────────────────────────────
  AirportData(
    icao: 'OMDB', iata: 'DXB', name: 'Dubai International Airport',
    city: 'Dubai', country: 'UAE', countryCode: 'AE', flag: '\u{1F1E6}\u{1F1EA}',
    lat: 25.2528, lng: 55.3644,
    fr24Url: 'https://www.flightradar24.com/airport/dxb',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OMDB',
  ),
  AirportData(
    icao: 'OMAA', iata: 'AUH', name: 'Abu Dhabi International Airport',
    city: 'Abu Dhabi', country: 'UAE', countryCode: 'AE', flag: '\u{1F1E6}\u{1F1EA}',
    lat: 24.4330, lng: 54.6511,
    fr24Url: 'https://www.flightradar24.com/airport/auh',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OMAA',
  ),
  AirportData(
    icao: 'OMSJ', iata: 'SHJ', name: 'Sharjah International Airport',
    city: 'Sharjah', country: 'UAE', countryCode: 'AE', flag: '\u{1F1E6}\u{1F1EA}',
    lat: 25.3286, lng: 55.5172,
    fr24Url: 'https://www.flightradar24.com/airport/shj',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OMSJ',
  ),
  AirportData(
    icao: 'OMDW', iata: 'DWC', name: 'Al Maktoum International Airport',
    city: 'Dubai', country: 'UAE', countryCode: 'AE', flag: '\u{1F1E6}\u{1F1EA}',
    lat: 24.8960, lng: 55.1614,
    fr24Url: 'https://www.flightradar24.com/airport/dwc',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OMDW',
  ),
  AirportData(
    icao: 'OMRK', iata: 'RKT', name: 'Ras Al Khaimah International Airport',
    city: 'Ras Al Khaimah', country: 'UAE', countryCode: 'AE', flag: '\u{1F1E6}\u{1F1EA}',
    lat: 25.6135, lng: 55.9388,
    fr24Url: 'https://www.flightradar24.com/airport/rkt',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OMRK',
  ),

  // ── KSA ─────────────────────────────────────────────────────────
  AirportData(
    icao: 'OEJN', iata: 'JED', name: 'King Abdulaziz International Airport',
    city: 'Jeddah', country: 'Saudi Arabia', countryCode: 'SA', flag: '\u{1F1F8}\u{1F1E6}',
    lat: 21.6796, lng: 39.1565,
    fr24Url: 'https://www.flightradar24.com/airport/jed',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OEJN',
  ),
  AirportData(
    icao: 'OERK', iata: 'RUH', name: 'King Khalid International Airport',
    city: 'Riyadh', country: 'Saudi Arabia', countryCode: 'SA', flag: '\u{1F1F8}\u{1F1E6}',
    lat: 24.9576, lng: 46.6988,
    fr24Url: 'https://www.flightradar24.com/airport/ruh',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OERK',
  ),
  AirportData(
    icao: 'OEDF', iata: 'DMM', name: 'King Fahd International Airport',
    city: 'Dammam', country: 'Saudi Arabia', countryCode: 'SA', flag: '\u{1F1F8}\u{1F1E6}',
    lat: 26.4712, lng: 49.7979,
    fr24Url: 'https://www.flightradar24.com/airport/dmm',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OEDF',
  ),
  AirportData(
    icao: 'OEMA', iata: 'MED', name: 'Prince Mohammed bin Abdulaziz Airport',
    city: 'Madinah', country: 'Saudi Arabia', countryCode: 'SA', flag: '\u{1F1F8}\u{1F1E6}',
    lat: 24.5534, lng: 39.7051,
    fr24Url: 'https://www.flightradar24.com/airport/med',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OEMA',
  ),

  // ── OMAN ────────────────────────────────────────────────────────
  AirportData(
    icao: 'OOMS', iata: 'MCT', name: 'Muscat International Airport',
    city: 'Muscat', country: 'Oman', countryCode: 'OM', flag: '\u{1F1F4}\u{1F1F2}',
    lat: 23.5933, lng: 58.2844,
    fr24Url: 'https://www.flightradar24.com/airport/mct',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OOMS',
  ),
  AirportData(
    icao: 'OOSA', iata: 'SLL', name: 'Salalah Airport',
    city: 'Salalah', country: 'Oman', countryCode: 'OM', flag: '\u{1F1F4}\u{1F1F2}',
    lat: 17.0387, lng: 54.0913,
    fr24Url: 'https://www.flightradar24.com/airport/sll',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OOSA',
  ),

  // ── QATAR ───────────────────────────────────────────────────────
  AirportData(
    icao: 'OTHH', iata: 'DOH', name: 'Hamad International Airport',
    city: 'Doha', country: 'Qatar', countryCode: 'QA', flag: '\u{1F1F6}\u{1F1E6}',
    lat: 25.2731, lng: 51.6081,
    fr24Url: 'https://www.flightradar24.com/airport/doh',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OTHH',
  ),

  // ── BAHRAIN ─────────────────────────────────────────────────────
  AirportData(
    icao: 'OBBI', iata: 'BAH', name: 'Bahrain International Airport',
    city: 'Manama', country: 'Bahrain', countryCode: 'BH', flag: '\u{1F1E7}\u{1F1ED}',
    lat: 26.2708, lng: 50.6336,
    fr24Url: 'https://www.flightradar24.com/airport/bah',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OBBI',
  ),

  // ── ISRAEL ──────────────────────────────────────────────────────
  AirportData(
    icao: 'LLBG', iata: 'TLV', name: 'Ben Gurion International Airport',
    city: 'Tel Aviv', country: 'Israel', countryCode: 'IL', flag: '\u{1F1EE}\u{1F1F1}',
    lat: 32.0114, lng: 34.8867,
    fr24Url: 'https://www.flightradar24.com/airport/tlv',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=LLBG',
  ),
  AirportData(
    icao: 'LLER', iata: 'ETH', name: 'Ramon International Airport',
    city: 'Eilat', country: 'Israel', countryCode: 'IL', flag: '\u{1F1EE}\u{1F1F1}',
    lat: 29.7270, lng: 35.0124,
    fr24Url: 'https://www.flightradar24.com/airport/eth',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=LLER',
  ),

  // ── LEBANON ─────────────────────────────────────────────────────
  AirportData(
    icao: 'OLBA', iata: 'BEY', name: 'Rafic Hariri International Airport',
    city: 'Beirut', country: 'Lebanon', countryCode: 'LB', flag: '\u{1F1F1}\u{1F1E7}',
    lat: 33.8209, lng: 35.4884,
    fr24Url: 'https://www.flightradar24.com/airport/bey',
    notamUrl: 'https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=OLBA',
  ),
];

File: backend/src/routes/airports.mjs
// =============================================================================
// BRE4CH v2 - Airport Status API
// Real-time airport status from ICAO NOTAMs + FAA TFRS + headline analysis
//
// GET /api/airports/status  → all airports with live status
// =============================================================================

import { Router } from 'express';

const router = Router();

// ── Regional airports (ICAO codes) ──────────────────────────────

const AIRPORTS = [
  // UAE
  { icao: 'OMDB', iata: 'DXB', name: 'Dubai Intl', country: 'AE' },
  { icao: 'OMAA', iata: 'AUH', name: 'Abu Dhabi Intl', country: 'AE' },
  { icao: 'OMSJ', iata: 'SHJ', name: 'Sharjah Intl', country: 'AE' },
  { icao: 'OMDW', iata: 'DWC', name: 'Al Maktoum Intl', country: 'AE' },
  { icao: 'OMRK', iata: 'RKT', name: 'RAK Intl', country: 'AE' },
  // KSA
  { icao: 'OEJN', iata: 'JED', name: 'King Abdulaziz Intl', country: 'SA' },
  { icao: 'OERK', iata: 'RUH', name: 'King Khalid Intl', country: 'SA' },
  { icao: 'OEDF', iata: 'DMM', name: 'King Fahd Intl', country: 'SA' },
  { icao: 'OEMA', iata: 'MED', name: 'Madinah Intl', country: 'SA' },
  // Oman
  { icao: 'OOMS', iata: 'MCT', name: 'Muscat Intl', country: 'OM' },
  { icao: 'OOSA', iata: 'SLL', name: 'Salalah', country: 'OM' },
  // Qatar
  { icao: 'OTHH', iata: 'DOH', name: 'Hamad Intl', country: 'QA' },
  // Bahrain
  { icao: 'OBBI', iata: 'BAH', name: 'Bahrain Intl', country: 'BH' },
  // Israel
  { icao: 'LLBG', iata: 'TLV', name: 'Ben Gurion Intl', country: 'IL' },
  { icao: 'LLER', iata: 'ETH', name: 'Ramon Intl', country: 'IL' },
  // Lebanon
  { icao: 'OLBA', iata: 'BEY', name: 'Rafic Hariri Intl', country: 'LB' },
];

// ── NOTAM keywords that indicate closures/restrictions ──────────

const CLOSURE_KEYWORDS = [
  'CLSD', 'CLOSED', 'AD CLSD', 'RWY CLSD',
  'APRN CLSD', 'TWY CLSD', 'NOT AVBL',
];

const RESTRICTION_KEYWORDS = [
  'RESTRICTED', 'AIRSPACE CLOSED', 'NOTAM',
  'TFR', 'FIR CLSD', 'CTR CLSD',
  'MILITARY ACTIVITY', 'DRONE ACTIVITY',
  'LIMITED', 'DIVERSION', 'ALTERNATE',
];

const GROUND_STOP_KEYWORDS = [
  'GROUND STOP', 'GDP', 'GROUND DELAY',
  'FLOW CONTROL', 'SLOT RESTRICTION',
];

// ── Cache ────────────────────────────────────────────────────────

let cachedStatus = null;
let cacheTime = 0;
const CACHE_TTL = 90_000; // 90s

// ── Fetch NOTAMs from FAA DINS (public, no auth) ────────────────

async function fetchNotamsForAirport(icao) {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10_000);
    const url = `https://www.notams.faa.gov/dinsQueryWeb/queryRetrievalMapAction.do?reportType=Raw&retrieveLocId=${icao}&actionType=notamRetrievalByICAOs`;
    const res = await fetch(url, {
      headers: { 'User-Agent': 'BRE4CH-ROAR/1.0' },
      signal: controller.signal,
    });
    clearTimeout(timeout);

    if (!res.ok) return [];
    const text = await res.text();

    // Extract NOTAM text blocks
    const notams = [];
    const regex = /NOTAM\s*\d*[^]*?(?=NOTAM\s*\d|$)/gi;
    let match;
    while ((match = regex.exec(text)) !== null) {
      notams.push(match[0].trim());
    }
    return notams;
  } catch {
    return [];
  }
}

// ── Analyze NOTAMs to determine airport status ──────────────────

function analyzeStatus(notams, icao) {
  const allText = notams.join(' ').toUpperCase();

  // Check for full closure
  for (const kw of CLOSURE_KEYWORDS) {
    if (allText.includes(kw)) {
      return {
        status: 'CLOSED',
        traffic: 'SUSPENDED',
        notamCount: notams.length,
        reason: `NOTAM: ${kw} detected`,
      };
    }
  }

  // Check for ground stops / delays
  for (const kw of GROUND_STOP_KEYWORDS) {
    if (allText.includes(kw)) {
      return {
        status: 'RESTRICTED',
        traffic: 'DELAYED',
        notamCount: notams.length,
        reason: `NOTAM: ${kw}`,
      };
    }
  }

  // Check for restrictions
  let restrictionCount = 0;
  for (const kw of RESTRICTION_KEYWORDS) {
    if (allText.includes(kw)) restrictionCount++;
  }

  if (restrictionCount >= 3) {
    return {
      status: 'RESTRICTED',
      traffic: 'DISRUPTED',
      notamCount: notams.length,
      reason: `${restrictionCount} restriction NOTAMs active`,
    };
  }

  if (notams.length > 5) {
    return {
      status: 'OPEN',
      traffic: 'CAUTION',
      notamCount: notams.length,
      reason: `${notams.length} active NOTAMs — elevated activity`,
    };
  }

  return {
    status: 'OPEN',
    traffic: 'NORMAL',
    notamCount: notams.length,
    reason: notams.length > 0 ? `${notams.length} routine NOTAMs` : 'No active NOTAMs',
  };
}

// ── Headline-based override (from cached headlines) ─────────────

function headlineOverrides(headlines) {
  const overrides = {};
  if (!headlines || !headlines.length) return overrides;

  for (const h of headlines) {
    const title = (h.title || '').toLowerCase();

    // Dubai airport
    if ((title.includes('dubai') || title.includes('dxb')) &&
        (title.includes('airport') || title.includes('closed') || title.includes('shut') || title.includes('suspended'))) {
      if (title.includes('closed') || title.includes('shut') || title.includes('suspended')) {
        overrides['OMDB'] = { status: 'CLOSED', traffic: 'SUSPENDED', reason: `HEADLINE: ${h.title}`, source: h.source };
      } else if (title.includes('divert') || title.includes('delay')) {
        overrides['OMDB'] = { status: 'RESTRICTED', traffic: 'DISRUPTED', reason: `HEADLINE: ${h.title}`, source: h.source };
      }
    }

    // Ben Gurion
    if ((title.includes('ben gurion') || title.includes('tlv') || title.includes('tel aviv airport')) &&
        (title.includes('closed') || title.includes('shut') || title.includes('suspended') || title.includes('divert'))) {
      overrides['LLBG'] = { status: 'CLOSED', traffic: 'SUSPENDED', reason: `HEADLINE: ${h.title}`, source: h.source };
    }

    // Beirut
    if ((title.includes('beirut') || title.includes('rafic hariri') || title.includes('bey')) &&
        (title.includes('closed') || title.includes('shut') || title.includes('strike') || title.includes('damaged'))) {
      overrides['OLBA'] = { status: 'CLOSED', traffic: 'SUSPENDED', reason: `HEADLINE: ${h.title}`, source: h.source };
    }

    // Generic airspace closure
    if (title.includes('airspace') && title.includes('closed')) {
      if (title.includes('iran')) {
        // No Iranian airports in our list, but note it
      }
      if (title.includes('israel')) {
        overrides['LLBG'] = overrides['LLBG'] || { status: 'RESTRICTED', traffic: 'DISRUPTED', reason: `HEADLINE: ${h.title}`, source: h.source };
        overrides['LLER'] = overrides['LLER'] || { status: 'RESTRICTED', traffic: 'DISRUPTED', reason: `HEADLINE: ${h.title}`, source: h.source };
      }
      if (title.includes('lebanon')) {
        overrides['OLBA'] = overrides['OLBA'] || { status: 'RESTRICTED', traffic: 'DISRUPTED', reason: `HEADLINE: ${h.title}`, source: h.source };
      }
    }
  }

  return overrides;
}

// ── Main endpoint ───────────────────────────────────────────────

router.get('/airports/status', async (req, res) => {
  const now = Date.now();

  // Return cache if fresh
  if (cachedStatus && (now - cacheTime) < CACHE_TTL) {
    return res.json({ ...cachedStatus, cached: true });
  }

  try {
    // Fetch NOTAMs for all airports in parallel
    const notamResults = await Promise.allSettled(
      AIRPORTS.map(async (ap) => {
        const notams = await fetchNotamsForAirport(ap.icao);
        const analysis = analyzeStatus(notams, ap.icao);
        return { ...ap, ...analysis };
      })
    );

    // Get headline overrides (from sources route if available)
    let headlines = [];
    try {
      // Import cached headlines from sources module
      const sourcesModule = await import('./sources.mjs');
      if (sourcesModule.getCachedHeadlines) {
        headlines = sourcesModule.getCachedHeadlines();
      }
    } catch { /* sources not available */ }

    const overrides = headlineOverrides(headlines);

    // Build results
    const airports = notamResults.map((r) => {
      if (r.status === 'fulfilled') {
        const ap = r.value;
        // Apply headline overrides if they indicate worse status
        if (overrides[ap.icao]) {
          const ov = overrides[ap.icao];
          // Only override if headline indicates worse status
          if (ov.status === 'CLOSED' || (ov.status === 'RESTRICTED' && ap.status === 'OPEN')) {
            return { ...ap, ...ov, notamCount: ap.notamCount, headlineOverride: true };
          }
        }
        return ap;
      }
      // Fallback for failed requests
      const ap = AIRPORTS.find(a => true); // shouldn't happen
      return { ...ap, status: 'UNKNOWN', traffic: 'UNKNOWN', notamCount: 0, reason: 'NOTAM fetch failed' };
    });

    const result = {
      airports,
      count: airports.length,
      lastRefresh: new Date().toISOString(),
      open: airports.filter(a => a.status === 'OPEN').length,
      restricted: airports.filter(a => a.status === 'RESTRICTED').length,
      closed: airports.filter(a => a.status === 'CLOSED').length,
      cached: false,
    };

    cachedStatus = result;
    cacheTime = now;
    res.json(result);
  } catch (err) {
    console.error(`[AIRPORTS] Error: ${err.message}`);
    // Return last cache or empty
    if (cachedStatus) {
      return res.json({ ...cachedStatus, cached: true, stale: true });
    }
    res.status(500).json({ error: 'Airport status fetch failed' });
  }
});

export default router;
